### PR TITLE
Fixed ClippingOP.getNewAlignmentStartOffset() to handle N cases properly and added unit tests

### LINKED ISSUE #2055: ClippingOP incorrectly changes reference start position when softclipping reads with N in cigar strings
I have discovered that when I run `ClippingOP.apply(SOFTCLIP_BASES, read)` for a read with the cigar string `1S3N2M10N1M4H` with clippingOP.start = 12 & clippingOP.stop = 12 it will correctly softclip the read to have `1S3N2M1S4H` as its new cigar string. Unfortunately this new read will have incremented the starting position for the new read by 3, (ex. startPosition=1000 -> startPostion=1003). This appears to be an issue with `ClippingOP.getNewAlignmentStartOffset()` and the fact that N consumes reference bases but not read bases.

### PR DESCRIPTION
**Pull request by jamesemery**
resolves #2055

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 1]
```diff
@@ -178,20 +178,26 @@ private GATKRead applyREVERT_SOFTCLIPPED_BASES(final GATKRead read) {
     /**
      * Given a cigar string, get the number of bases hard or soft clipped at the start
      */
-    private int getNewAlignmentStartOffset(final Cigar __cigar, final Cigar __oldCigar) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: Add `@VisibleForTesting`

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 2]
```diff
@@ -178,20 +178,26 @@ private GATKRead applyREVERT_SOFTCLIPPED_BASES(final GATKRead read) {
...
+    static int getNewAlignmentStartOffset(final Cigar __cigar, final Cigar __oldCigar) {
         int num = 0;
+
+        int oldNum = 0;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: Needs a better name -- `oldNum` is not informative.

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: Also document the semantics for this (eg., can it temporarily become negative)?

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 3]
```diff
@@ -219,7 +225,7 @@ private int getNewAlignmentStartOffset(final Cigar __cigar, final Cigar __oldCig
             }
         }
 
-        return oldNum;
+        return Math.abs(oldNum); // if oldNum is negative it means some of the preceeding N/Ds were trimmed but not all, should be fixed
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: Comment needs adjusting -- "should be fixed" implies a new ticket

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 4]
```diff
@@ -178,20 +178,26 @@ private GATKRead applyREVERT_SOFTCLIPPED_BASES(final GATKRead read) {
     /**
      * Given a cigar string, get the number of bases hard or soft clipped at the start
      */
-    private int getNewAlignmentStartOffset(final Cigar __cigar, final Cigar __oldCigar) {
+    static int getNewAlignmentStartOffset(final Cigar __cigar, final Cigar __oldCigar) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: get rid of strange underscore naming scheme

File: src/test/java/org/broadinstitute/hellbender/utils/clipping/ClippingOpUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        Cigar c2BackClipped = TextCigarCodec.decode("30M30S");
+        Cigar c2FrontClippedTrivial = TextCigarCodec.decode("15S15M10N30M");
+        Cigar c2BackClippedTrivial = TextCigarCodec.decode("30M10N15M15S");
+        Assert.assertEquals(ClippingOp.getNewAlignmentStartOffset(c2FrontClippedImproper, c2),  30, "c2FrontClippedImprope");
+        Assert.assertEquals(ClippingOp.getNewAlignmentStartOffset(c2FrontClippedPartial, c2),  35, "c2FrontClippedPartial");
+        Assert.assertEquals(ClippingOp.getNewAlignmentStartOffset(c2BackClippedImproper, c2),  0, "c2BackClippedImproper");
+        Assert.assertEquals(ClippingOp.getNewAlignmentStartOffset(c2FrontClipped, c2),  40, "c2FrontClipped");
+        Assert.assertEquals(ClippingOp.getNewAlignmentStartOffset(c2BackClipped, c2),  0, "c2BackClipped");
+        Assert.assertEquals(ClippingOp.getNewAlignmentStartOffset(c2FrontClippedTrivial, c2),  15, "c2FrontClippedTrivial");
+        Assert.assertEquals(ClippingOp.getNewAlignmentStartOffset(c2BackClippedTrivial, c2),  0, "c2BackClippedTrivial");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/clipping/ClippingOpUnitTest.java]: Can you refactor to use a `DataProvider`, with each test case consisting of a `Cigar` plus expected new alignment start offset

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 6]
```diff
@@ -178,20 +178,26 @@ private GATKRead applyREVERT_SOFTCLIPPED_BASES(final GATKRead read) {
     /**
      * Given a cigar string, get the number of bases hard or soft clipped at the start
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: Is this method description accurate? If not, write new docs explaining what the return value means.

File: src/test/java/org/broadinstitute/hellbender/utils/clipping/ClippingOpUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        Cigar c4FrontClipped = TextCigarCodec.decode("10S20M10N10M");
+        Cigar c4BackClipped = TextCigarCodec.decode("10S10N20M10S");
+        Assert.assertEquals(ClippingOp.getNewAlignmentStartOffset(c4FrontClipped, c4),  10, "c4FrontClipped");
+        Assert.assertEquals(ClippingOp.getNewAlignmentStartOffset(c4BackClipped, c4),  0, "c4BackClipped");
+
+        // Testing cases where readbses but not reference bases are taken
+        Cigar c5 = TextCigarCodec.decode("10S10I20M10I10M");
+        Cigar c5FrontClipped = TextCigarCodec.decode("20S20M10I10M");
+        Cigar c5BackClipped = TextCigarCodec.decode("10S10I20M20S");
+        Cigar c5PartiallyClipped = TextCigarCodec.decode("15S5I20M10I10M");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/clipping/ClippingOpUnitTest.java]: Do you need test cases that start with a hard clip?

Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/clipping/ClippingOpUnitTest.java]: In general, ensure that you have test cases for all possible 2-element starts.

**Comment by droazen** [comment]
Review complete -- back to @jamesemery

**Comment by jamesemery** [comment]
@droazen I responded to your comments and made the changes you recommended

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 8]
```diff
@@ -176,34 +176,44 @@ private GATKRead applyREVERT_SOFTCLIPPED_BASES(final GATKRead read) {
     }
 
     /**
-     * Given a cigar string, get the number of bases hard or soft clipped at the start
+     * Given two cigar strings corresponding to the a read that has been softcliped and before, returns an integer
+     * corresponding to the number of reference bases that the new string must be offset by in order to have the
+     * correct start according to the reference.
+     *
+     * @VisibleForTesting
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: `@VisibleForTesting` should be outside of the comment, and before the method header

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 9]
```diff
@@ -176,34 +176,44 @@ private GATKRead applyREVERT_SOFTCLIPPED_BASES(final GATKRead read) {
     }
 
     /**
-     * Given a cigar string, get the number of bases hard or soft clipped at the start
+     * Given two cigar strings corresponding to the a read that has been softcliped and before, returns an integer
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: `corresponding to the a read that has been softcliped and before` -> `from a read before and after soft-clipping`

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 10]
```diff
@@ -176,34 +176,44 @@ private GATKRead applyREVERT_SOFTCLIPPED_BASES(final GATKRead read) {
...
+     * corresponding to the number of reference bases that the new string must be offset by in order to have the
+     * correct start according to the reference.
+     *
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: Add `@param` javadoc for the method arguments

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 11]
```diff
@@ -176,34 +176,44 @@ private GATKRead applyREVERT_SOFTCLIPPED_BASES(final GATKRead read) {
...
+     * @VisibleForTesting
      */
-    private int getNewAlignmentStartOffset(final Cigar __cigar, final Cigar __oldCigar) {
-        int num = 0;
-        for (final CigarElement e : __cigar.getCigarElements()) {
+    static int getNewAlignmentStartOffset(final Cigar clippedCigar, final Cigar oldCigar) {
+        int readBasesBeforeReference = 0;
+
+        int basesBeforeReferenceOld = 0;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: Add a one-line comment for each of these variables explaining what it represents

**Comment by jamesemery** [comment]
@droazen I responded to your second round of comments

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 12]
```diff
@@ -176,34 +177,46 @@ private GATKRead applyREVERT_SOFTCLIPPED_BASES(final GATKRead read) {
     }
 
     /**
-     * Given a cigar string, get the number of bases hard or soft clipped at the start
+     * Given two cigar strings corresponding to the a read from a read before and after soft-clipping, returns an integer
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: `to the a read from a read` -> `to a read`

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 13]
```diff
@@ -176,34 +177,46 @@ private GATKRead applyREVERT_SOFTCLIPPED_BASES(final GATKRead read) {
...
+     *
+     * @param clippedCigar the new cigar string after clipping
+     * @param oldCigar the cigar string of the read before it was soft clipped
      */
-    private int getNewAlignmentStartOffset(final Cigar __cigar, final Cigar __oldCigar) {
-        int num = 0;
-        for (final CigarElement e : __cigar.getCigarElements()) {
+    @VisibleForTesting
+    static int getNewAlignmentStartOffset(final Cigar clippedCigar, final Cigar oldCigar) {
+        int readBasesBeforeReference = 0; // The number of reference bases consumed on the old cigar before read bases are consumed
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: Hmm, is this name accurate? The description says "The number of reference bases consumed on the old cigar before read bases are consumed", but the name implies the opposite (read bases before reference bases are consumed). Which is it?

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 14]
```diff
@@ -176,34 +177,46 @@ private GATKRead applyREVERT_SOFTCLIPPED_BASES(final GATKRead read) {
...
+
+        int basesBeforeReferenceOld = 0; // The number of reference bases consumed on the new cigar before read bases are consumed
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: Comment for `basesBeforeReferenceOld` says that it applies to "the new cigar", but the name implies that it applies to the old cigar. Which is it? (same comment applies to `readBasesBeforeReference` above)

File: src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java [hunk 15]
```diff
@@ -212,14 +225,14 @@ private int getNewAlignmentStartOffset(final Cigar __cigar, final Cigar __oldCig
...
+            basesBeforeReferenceOld += curRefLength;
 
-            if (curReadCounter > num || truncated) {
+            if (curReadCounter > readBasesBeforeReference || truncated) {
                 break;
             }
         }
 
-        return oldNum;
+        return Math.abs(basesBeforeReferenceOld); // if oldNum is negative it means some of the preceeding N/Ds were trimmed but not all so we take absolute value
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java]: `preceeding` -> `preceding`

File: src/test/java/org/broadinstitute/hellbender/utils/clipping/ClippingOpUnitTest.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+        Cigar cpostClip = TextCigarCodec.decode(postClip);
+        Assert.assertEquals(ClippingOp.getNewAlignmentStartOffset(cpostClip, cpreClip), result,
+                "getNewAlignmentStartOffset returned "+ClippingOp.getNewAlignmentStartOffset(cpostClip, cpreClip)+
+                        " when "+result+" was expected for "+preClip.toString()+" which was clipped to "+postClip.toString());
+    }
+
+    @DataProvider(name = "SoftClipedReadsNewStart")
+    public Object[][] makeRevertSoftClipsBeforeContig() {
+        return new Object[][] {
+                {"70M", "10S60M", 10},
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/clipping/ClippingOpUnitTest.java]: Add a one-line comment at the top of your `DataProvider` explaining what each of the three values here mean.

File: src/test/java/org/broadinstitute/hellbender/utils/clipping/ClippingOpUnitTest.java [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+import htsjdk.samtools.TextCigarCodec;
+import org.broadinstitute.hellbender.utils.test.BaseTest;
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+public final class ClippingOpUnitTest extends BaseTest {
+
+    @Test (dataProvider = "SoftClipedReadsNewStart")
+    public void testGetNewAlignmentStartOffset(final String preClip, final String postClip, final int result) {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/clipping/ClippingOpUnitTest.java]: `result` -> `expectedResult`

**Comment by droazen** [comment]
Final review complete, back to @jamesemery. Please rebase and squash into a single commit after addressing remaining comments.

**Comment by droazen** [comment]
@jamesemery Don't forget to click "Delete branch" to delete your branch after merging.

**Pull request merged**